test/gtest/ucp/test_ucp_worker.cc

```diff
@@ -0,0 +1,316 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "ucp_test.h"
+#include <uct/api/uct.h>
+#include <uct/api/tl.h>
+
+extern "C" {
+#include <ucp/core/ucp_worker.h>
+#include <ucp/wireup/wireup_ep.h>
+#include <uct/base/uct_iface.h>
+}
+
+
+class test_ucp_worker_discard : public ucp_test {
+public:
+    static ucp_params_t get_ctx_params() {
+        ucp_params_t params = ucp_test::get_ctx_params();
+        params.features |= UCP_FEATURE_TAG;
+        return params;
+    }
+
+protected:
+    void init() {
+        ucp_test::init();
+        m_created_ep_count     = 0;
+        m_destroyed_ep_count   = 0;
+        m_flush_ep_count       = 0;
+        m_pending_add_ep_count = 0;
+
+        m_flush_comps.clear();
+        m_pending_reqs.clear();
+    }
+
+    void add_pending_reqs(uct_ep_h uct_ep,
+                          uct_pending_callback_t func,
+                          std::vector<uct_pending_req_t> &pending_reqs,
+                          unsigned base = 0) {
+        for (unsigned i = 0; i < m_pending_purge_reqs_count; i++) {
+            pending_reqs[i].func = func;
+            uct_ep_pending_add(uct_ep, &pending_reqs[i], 0);
+        }
+    }
+
+    void test_worker_discard(void *ep_flush_func,
+                             void *ep_pending_add_func,
+                             void *ep_pending_purge_func,
+                             unsigned ep_count = 8,
+                             unsigned wireup_ep_count = 0,
+                             unsigned wireup_aux_ep_count = 0) {
+        uct_iface_ops_t ops                  = {0};
+        unsigned created_wireup_aux_ep_count = 0;
+        unsigned total_ep_count              = ep_count + wireup_aux_ep_count;
+        uct_iface_t iface;
+        std::vector<uct_ep_t> eps(total_ep_count);
+        std::vector<uct_ep_h> wireup_eps(wireup_ep_count);
+        ucp_ep_t ucp_ep;
+        ucs_status_t status;
+
+        ASSERT_LE(wireup_ep_count, ep_count);
+        ASSERT_LE(wireup_aux_ep_count, wireup_ep_count);
+
+        ucp_ep.worker = sender().worker();
+
+        ops.ep_flush         = (uct_ep_flush_func_t)ep_flush_func;
+        ops.ep_pending_add   = (uct_ep_pending_add_func_t)ep_pending_add_func;
+        ops.ep_pending_purge = (uct_ep_pending_purge_func_t)ep_pending_purge_func;
+        ops.ep_destroy       = ep_destroy_func;
+        iface.ops            = ops;
+
+        for (unsigned i = 0; i < ep_count; i++) {
+            uct_ep_h discard_ep;
+
+            eps[i].iface = &iface;
+            m_created_ep_count++;
+
+            unsigned expected_pending_purge_reqs_count =
+                (ep_pending_purge_func == ep_pending_purge_func_iter) ?
```

Comment:
minor: quite hard to read, i'd use `if/else` and add general comment (not for every line)

Reply:
fixed